io/StacReader.cpp

```diff
@@ -0,0 +1,521 @@
+        initializeCatalog(stacJson);
+    else
+        throw pdal_error("Could not initialize STAC object of type " + stacType);
+}
+
+void schemaFetch(const nlohmann::json_uri &json_uri, nlohmann::json &json)
+{
+    std::unique_ptr<arbiter::Arbiter> fetcher;
+    fetcher.reset(new arbiter::Arbiter());
+    std::string jsonStr = fetcher->get(json_uri.url());
+    json = nlohmann::json::parse(jsonStr);
+}
+
+void StacReader::schemaValidate(NL::json stacJson)
+{
+    std::function<void(const nlohmann::json_uri &, nlohmann::json &)> fetch = schemaFetch;
+    nlohmann::json_schema::json_validator val(
+        fetch,
+        [](const std::string &, const std::string &) {}
+    );
+    if (!stacJson.contains("type"))
+        throw pdal_error("Invalid STAC json");
+    std::string type = stacJson["type"].get<std::string>();
+    std::string schemaUrl;
+
+    if (type == "Feature")
+    {
+        schemaUrl = "https://schemas.stacspec.org/v1.0.0/item-spec/json-schema/item.json";
+        for (auto& extSchemaUrl: stacJson["stac_extensions"])
+        {
+            log()->get(LogLevel::Debug) << "Processing extension " << extSchemaUrl << std::endl;
+            std::string schemaStr = m_arbiter->get(extSchemaUrl);
+            NL::json schemaJson = NL::json::parse(schemaStr);
+            val.set_root_schema(schemaJson);
+            val.validate(stacJson);
+        }
+    }
+    else if (type == "Catalog")
+        schemaUrl = "https://schemas.stacspec.org/v1.0.0/catalog-spec/json-schema/catalog.json";
+    else
+        throw pdal_error("Invalid STAC type for PDAL consumption");
+
+    std::string schemaStr = m_arbiter->get(schemaUrl);
+    NL::json schemaJson = NL::json::parse(schemaStr);
+    val.set_root_schema(schemaJson);
+    val.validate(stacJson);
+}
+
+void StacReader::initializeItem(NL::json stacJson)
+{
+    if (prune(stacJson))
+        return;
+
+    if (m_args->schemaValidate)
+        schemaValidate(stacJson);
+
+    if (!stacJson["assets"].contains(m_args->assetName))
+        throw pdal_error("asset_name("+m_args->assetName+") doesn't match STAC object.");
+
+    std::string dataUrl = stacJson["assets"][m_args->assetName]["href"].get<std::string>();
+    std::string driver = m_factory.inferReaderDriver(dataUrl);
+
+    log()->get(LogLevel::Debug) << "Using driver " << driver <<
+        " for file " << dataUrl << std::endl;
+
+    Stage *reader = PluginManager<Stage>::createObject(driver);
+
+    if (!reader)
+        throwError("Unable to create reader for file '" + dataUrl + "'.");
+
+    Options readerOptions;
+    // add reader options defined in reader args to their respective readers
+    if (m_readerArgs.contains(driver)) {
+        NL::json args = m_readerArgs[driver].get<NL::json>();
+        for (auto& arg : args.items()) {
+            readerOptions.add(arg.key(), arg.value());
+        }
+    }
+
+    readerOptions.add("filename", dataUrl);
+    reader->setOptions(readerOptions);
+
+    if (m_readerList.size() > 0)
```

Comment:
If the m_readerList size is 0, is that an error?

Reply:
Not in this spot, but maybe elsewhere. Here we're just making sure that there are readers in the list before applying the previous back to the current reader's input.